File: LINQPadKit/TileMap.cs
using LINQPad.Controls;
using LINQPad.Controls.Core;

namespace LINQPadKit;

public partial class TileMap : Div
{
    public static void Import()
    {
        KitUtil.Load("package", "dist", $"fabric@5.3.1.min.js");
        KitUtil.Load("TileMap", "dist", $"tile-map.css");
        KitUtil.Load("TileMap", "dist", $"tile-map.js");
    }

    public enum Shape
    {
        None = 0,
        Rectangle = 1,
        Hexagon = 2,
    }

    public string Instance { get; set; }

    private readonly Div _canvas;
    private readonly Div _bridge;
    private readonly string _shape;
    private readonly int _cellWidth;
    private readonly int _rows;
    private readonly int _cols;

    public TileMap(Shape shape, int cellWidth, int rows, int cols)
    {
        _shape = shape switch
        {
            Shape.Rectangle => "rectangle",
            Shape.Hexagon => "hexagon",
            _ => "",
        };
        _cellWidth = cellWidth;
        _rows = rows;
        _cols = cols;

        _canvas = new Div();
        Children.Add(_canvas);

        _bridge = new Div();
        _bridge.HtmlElement.AddEventListener("cell:mouseDown",
        [
            "x: args.detail.x",
            "y: args.detail.y",
        ], CellMouseDown);
        _bridge.HtmlElement.AddEventListener("cell:colorChange",
        [
            "x: args.detail.x",
            "y: args.detail.y",
            "color: args.detail.color",
        ], ColorChange);
        _bridge.HtmlElement.AddEventListener("cell:imageChange",
        [
            "x: args.detail.x",
            "y: args.detail.y",
            "image: args.detail.image",
        ], ImageChange);
        _bridge.HtmlElement.AddEventListener("ready",
        [
            "rows: args.detail.rows",
            "cols: args.detail.cols",
        ], Ready);
        Children.Add(_bridge);
    }

    public Cell this[int x, int y] => new(this, x, y);

    protected override void OnRendering(EventArgs e)
    {
        var instance = $"kit_{nameof(TileMap)}_{HtmlElement.ID}";
        Instance = instance;

        HtmlElement.InvokeScript(false, "eval", $"var {instance} = new {nameof(TileMap)}('{_shape}', {_cellWidth}, {_rows}, {_cols});");
        HtmlElement.InvokeScript(false, "eval", $"{Instance}.render('{_canvas.HtmlElement.ID}', '{_bridge.HtmlElement.ID}');");
    }

    public void SetColor(int x, int y, string color)
    {
        HtmlElement.InvokeScript(false, "eval", $@"{Instance}.setColor({x}, {y}, '{color}')");
    }

    public string? GetColor(int x, int y)
    {
        return HtmlElement.InvokeScript(true, "eval", $@"{Instance}.getColor({x}, {y})") as string;
    }

    public void SetImage(int x, int y, string image)
    {
        HtmlElement.InvokeScript(false, "eval", $@"{Instance}.setImage({x}, {y}, '{image}')");
    }

    public string? GetImage(int x, int y)
    {
        return HtmlElement.InvokeScript(true, "eval", $@"{Instance}.getImage({x}, {y})") as string;
    }

    public void ResetCell(int x, int y)
    {
        HtmlElement.InvokeScript(false, "eval", $@"{Instance}.resetCell({x}, {y})");
    }

    public struct CellMouseDownEventArgs
    {
        public int X { get; set; }
        public int Y { get; set; }
    }
    public delegate void CellMouseDownEvent(object sender, CellMouseDownEventArgs args);
    public event CellMouseDownEvent OnCellMouseDown;
    private void CellMouseDown(object? sender, PropertyEventArgs args)
    {
        OnCellMouseDown?.Invoke(this, new CellMouseDownEventArgs
        {
            X = int.Parse(args.Properties["x"]),
            Y = int.Parse(args.Properties["y"]),
        });
    }

    public struct CellColorChangeEventArgs
    {
        public int X { get; set; }
        public int Y { get; set; }
        public string Color { get; set; }
    }
    public delegate void CellColorChangeEvent(object sender, CellColorChangeEventArgs args);
    public event CellColorChangeEvent OnCellColorChange;
    private void ColorChange(object? sender, PropertyEventArgs args)
    {
        OnCellColorChange?.Invoke(this, new CellColorChangeEventArgs
        {
            X = int.Parse(args.Properties["x"]),
            Y = int.Parse(args.Properties["y"]),
            Color = args.Properties["color"],
        });
    }

    public struct CellImageChangeEventArgs
    {
        public int X { get; set; }
        public int Y { get; set; }
        public string Image { get; set; }
    }
    public delegate void CellImageChangeEvent(object sender, CellImageChangeEventArgs args);
    public event CellImageChangeEvent OnCellImageChange;
    private void ImageChange(object? sender, PropertyEventArgs args)
    {
        OnCellImageChange?.Invoke(this, new CellImageChangeEventArgs
        {
            X = int.Parse(args.Properties["x"]),
            Y = int.Parse(args.Properties["y"]),
            Image = args.Properties["image"],
        });
    }

    public struct ReadyEventArgs
    {
        public int Rows { get; set; }
        public int Cols { get; set; }
    }
    public delegate void ReadyEvent(object sender, ReadyEventArgs args);
    public event ReadyEvent OnReady;
    private void Ready(object? sender, PropertyEventArgs args)
    {
        OnReady?.Invoke(this, new ReadyEventArgs
        {
            Rows = int.Parse(args.Properties["rows"]),
            Cols = int.Parse(args.Properties["cols"]),
        });
    }
}

File: LINQPadKit.Dev/src/TileMap/tile-map.ts
/// <reference path="./fabric-impl.d.ts" />
type Shape = 'rectangle' | 'hexagon';

interface Cell {
    element: fabric.Object;
    x: number;
    y: number;
    color?: string;
    image?: string;
}

class TileMap {
    net_bridge?: string;
    canvas?: HTMLCanvasElement;
    fabricCanvas?: fabric.Canvas;
    cells: Cell[][] = [];
    cellHeight: number;

    onCellMouseDown: ((sender: TileMap, args: { x: number, y: number }) => void) | null = null;
    onCellColorChange: ((sender: TileMap, args: { color: string }) => void) | null = null;
    onImageChange: ((sender: TileMap, args: { image: string }) => void) | null = null;
    onReady: ((sender: TileMap, args: { rows: number, cols: number }) => void) | null = null;

    constructor(public shape: Shape, public cellWidth: number, public rows: number, public cols: number) {
        this.cellHeight = this.cellWidth / 2 * Math.sqrt(3) / 2 * 3;
    }

    render(id: string, net_bridge?: string) {
        this.net_bridge = net_bridge;

        const element = document.getElementById(id);
        if (element == null) throw "No element found.";

        element.innerHTML = '';
        const canvas = document.createElement('canvas');
        if (canvas == null) throw "No canvas found.";

        if (this.shape == 'rectangle') {
            canvas.width = this.cellWidth * this.cols;
            canvas.height = this.cellWidth * this.rows;
        }
        else if (this.shape == 'hexagon') {
            canvas.width = this.cellWidth * this.cols + this.cellWidth / 2;
            canvas.height = (this.cellWidth / 2 * Math.sqrt(3) / 2) * (this.rows * 2 + 1);
        }
        element.appendChild(canvas);
        this.canvas = canvas;
        this.fabricCanvas = new fabric.Canvas(canvas);
        this.fabricCanvas.selection = false;
        this.init(this.shape);
        this.ready({ rows: this.rows, cols: this.cols });
    }

    private init(shape) {
        if (this.fabricCanvas == null) throw "No fabric canvas found."

        this.cells = [];
        if (shape == 'rectangle') {
            for (let y = 0; y < this.rows; y++) {
                const row: Cell[] = [];
                for (let x = 0; x < this.cols; x++) {
                    const rect = new fabric.Rect({
                        left: x * this.cellWidth,
                        top: y * this.cellWidth,
                        fill: this.getOriginColor(y, x),
                        width: this.cellWidth,
                        height: this.cellWidth,
                        selectable: false,
                        hoverCursor: 'default'
                    });
                    const cell: Cell = { element: rect, y: y, x: x };
                    rect.on('mousedown', e => {
                        this.mouseDown({ x: cell.x, y: cell.y });
                    });
                    this.fabricCanvas.add(rect);
                    row.push(cell);
                }
                this.cells.push(row);
            }
        }
        else if (shape == 'hexagon') {
            for (let y = 0; y < this.rows; y++) {
                const row: Cell[] = [];
                for (let x = 0; x < this.cols; x++) {
                    const height = this.cellWidth / 2 * Math.sqrt(3) / 2;
                    const left = x * this.cellWidth + (y % 2 == 0 ? 0 : this.cellWidth / 2);
                    const top = y * this.cellWidth / 2 * Math.sqrt(3);
                    const points = [
                        [left + Math.floor(this.cellWidth / 2), top],
                        [left + this.cellWidth, top + height],
                        [left + this.cellWidth, top + height * 2],
                        [left + Math.floor(this.cellWidth / 2), top + height * 3],
                        [left, top + height * 2],
                        [left, top + height],
                    ];

                    const path = new fabric.Path(`M ${points.map(p => `${p[0]} ${p[1]}`).join(' L ')} z`);
                    path.fill = this.getOriginColor(y, x);
                    path.selectable = false;
                    path.hoverCursor = 'default';

                    const cell: Cell = {
                        element: path,
                        y, x,
                        color: path.fill,
                    };
                    path.on('mousedown', e => {
                        this.mouseDown({ x: cell.x, y: cell.y });
                    });
                    this.fabricCanvas.add(path);
                    row.push(cell);
                }
                this.cells.push(row);
            }
        }
    };

    setImage(x, y, image: string) {
        var img = new Image();
        img.src = image;
        img.onload = () => {
            if (this.fabricCanvas == null) throw "No fabric canvas found."

            var pattern = new fabric.Pattern({
                source: img,
                repeat: 'no-repeat',
                offsetX: (this.cellWidth / 2) - (img.width / 2),
                offsetY: (this.cellHeight / 2) - (img.height / 2)
            });

            const cell = this.cells[y][x];
            cell.color = undefined;
            cell.image = image;
            cell.element.set('fill', pattern);
            this.fabricCanvas.renderAll();
            this.imageChange?.({ x, y, image });
        };
    }

    getImage(x, y) {
        const cell = this.cells[y][x];
        return cell.image;
    }

    setColor(x, y, color) {
        if (this.fabricCanvas == null) throw "No fabric canvas found."

        const cell = this.cells[y][x];
        cell.image = undefined;
        cell.color = color;
        cell.element.set('fill', color);
        this.fabricCanvas.renderAll();
        this.colorChange?.({ x, y, color });
    }

    getColor(x, y) {
        const cell = this.cells[y][x];
        return cell.color;
    }

    resetCell(x, y) {
        const color = this.getOriginColor(x, y);
        this.setColor(x, y, color);
    }

    getOriginColor(x, y) {
        return '#C0C0C0';
    }

    private callBridge(name: string, detail) {
        if (this.net_bridge) {
            const element = document.getElementById(this.net_bridge);
            if (element == null) throw "No element found."
            const _event = new CustomEvent(name, { detail });
            element.dispatchEvent(_event);
        }
    }

    private mouseDown(detail: { x: number, y: number }) {
        this.onCellMouseDown?.(this, detail);
        this.callBridge('cell:mouseDown', detail);
    }

    private colorChange(detail: { x: number, y: number, color: string }) {
        this.onCellColorChange?.(this, detail);
        this.callBridge('cell:colorChange', detail);
    }

    private imageChange(detail: { x: number, y: number, image: string }) {
        this.onImageChange?.(this, detail);
        this.callBridge('cell:imageChange', detail);
    }

    private ready(detail: { rows: number, cols: number }) {
        this.onReady?.(this, detail);
        this.callBridge('ready', detail);
    }
}
